### PR TITLE
First DrugService Method

### PR DESCRIPTION
**Pull request by robertogagliotta**
Hi @MatthewVita 
This is Method is easy one, and i would like you check my code out before I move on. I did unit testing with the Drug Inventory -> delete and it was ok

### CODE REVIEW

**Comment by MatthewVita** [comment]
@robertogagliotta This looks awesome so far! Great work!

**Comment by MatthewVita** [comment]
BTW, we can fix up the Travis style complaints later. I would just focus on adding the other methods for now. Let me know if you need anything!

File: interface/drugs/add_edit_drug.php [hunk 1]
```diff
@@ -14,6 +14,8 @@
  require_once("drugs.inc.php");
  require_once("$srcdir/options.inc.php");
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: better to do the following (place this line below the require_once):
use OpenEMR\Services\DrugServices;

File: interface/drugs/add_edit_drug.php [hunk 2]
```diff
@@ -14,6 +14,8 @@
...
 
+$drugService = new \services\DrugService();
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: then this will be:
$drugService = new DrugService();

File: services/DrugService.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+class DrugService
+{
+   /**
+    * Default constructor
+    */
+   public function __construct()
+   {
+   }
+
+  public function delete_drug($drug_id)
```
Comment (bradymiller) [inline services/DrugService.php]: use camel case for PSR2 compliance:
``````
public function deleteDrug($drug_id)

File: services/DrugService.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
+<?php
+/**
+
+    DrugService
+
+*/
+
+namespace services;
```
Comment (bradymiller) [inline services/DrugService.php]: change to:
namespace OpenEMR\Services;

File: interface/drugs/add_edit_drug.php [hunk 5]
```diff
@@ -196,10 +198,7 @@ function sel_related() {
             sqlStatement("DELETE FROM drug_templates WHERE drug_id = ?", array($drug_id));
         } else { // deleting
             if (acl_check('admin', 'super')) {
-                sqlStatement("DELETE FROM drug_inventory WHERE drug_id = ?", array($drug_id));
-                sqlStatement("DELETE FROM drug_templates WHERE drug_id = ?", array($drug_id));
-                sqlStatement("DELETE FROM drugs WHERE drug_id = ?", array($drug_id));
-                sqlStatement("DELETE FROM prices WHERE pr_id = ? AND pr_selector != ''", array($drug_id));
+               $delete_drug_id = $drugService->delete_drug($drug_id);
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: camelcase the delete_drug to deleteDrug

File: interface/drugs/add_edit_drug.php (hunk 1 above)
Comment (bradymiller) [reply interface/drugs/add_edit_drug.php]: disregard above comment for now. I am in process of converting this namespace, so can do this in the future.

File: interface/drugs/add_edit_drug.php (hunk 2 above)
Comment (bradymiller) [reply interface/drugs/add_edit_drug.php]: disregard above comment for now. I am in process of converting this namespace, so can do this in the future.

File: services/DrugService.php (hunk 4 above)
Comment (bradymiller) [reply services/DrugService.php]: disregard above comment for now. I am in process of converting this namespace, so can do this in the future.

**Comment by bradymiller** [comment]
a new class gets the "happy class dance" :)
img_2681

**Comment by robertogagliotta** [comment]
@bradymiller @MatthewVita 
Previous review comments were solved and 3 more methods were added. 
I will keep sweeping sql code.

**Comment by bradymiller** [comment]
Hi @robertogagliotta ,
Looks like some PSR2 formatting issues to address.
Also, is this working on testing for you?
-brady

File: interface/drugs/add_edit_drug.php [hunk 6]
```diff
@@ -176,55 +182,50 @@ function sel_related() {
...
+            "max_level" => escapedff('form_max_level'),
+            "form" => escapedff('form_form'),
+            "size" => escapedff('form_sixe'),
+            "unit" => escapedff('form_unit'),
+            "route" => escapedff('form_route'),
+            "cyp_factor" => numericff('form_cyp_factor'),
+            "related_code" => escapedff('form_related_code'),
+            "allow_multiple" => (empty($_POST['form_allow_multiple' ]) ? 0 : 1),
+            "allow_combining" => (empty($_POST['form_allow_combining']) ? 0 : 1),
+            "active" => (empty($_POST['form_active']) ? 0 : 1)
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: remove all the escapedff (and change the numericff to a trim()) and instead use sql binding in the function in the service class.

File: interface/drugs/add_edit_drug.php [hunk 7]
```diff
@@ -176,55 +182,50 @@ function sel_related() {
...
-            sqlStatement("DELETE FROM drug_templates WHERE drug_id = ?", array($drug_id));
+           $updateDrug = array(
+            "name" => escapedff('form_name'),
+            "ndc_number" => escapedff('form_ndc_number'),
+            "drug_code" => escapedff('form_drug_code'),
+            "on_order" => escapedff('form_on_order'),
+            "reorder_point" => escapedff('form_reorder_point'),
+            "max_level" => escapedff('form_max_level'),
+            "form" => escapedff('form_form'),
+            "size" => escapedff('form_sixe'),
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: typo

File: interface/drugs/add_edit_drug.php [hunk 8]
```diff
@@ -176,55 +182,50 @@ function sel_related() {
...
+            "unit" => escapedff('form_unit'),
+            "route" => escapedff('form_route'),
+            "cyp_factor" => numericff('form_cyp_factor'),
+            "related_code" => escapedff('form_related_code'),
+            "allow_multiple" => (empty($_POST['form_allow_multiple' ]) ? 0 : 1),
+            "allow_combining" => (empty($_POST['form_allow_combining']) ? 0 : 1),
+            "active" => (empty($_POST['form_active']) ? 0 : 1)
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: remove all the escapedff (and change the numericff to a trim()) and instead use sql binding in the function in the service class.

File: interface/drugs/add_edit_drug.php [hunk 9]
```diff
@@ -235,7 +236,8 @@ function sel_related() {
             $tmpl["1"]['selector'] = $_POST['form_name'];
         }
 
-        sqlStatement("DELETE FROM prices WHERE pr_id = ? AND pr_selector != ''", array($drug_id));
+        $drugService->deletePrices($drug_id);
+        //sqlStatement("DELETE FROM prices WHERE pr_id = ? AND pr_selector != ''", array($drug_id));
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: remove instead of uncomment

File: services/DrugService.php [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+  public function deleteDrug($drug_id)
+  {
+    sqlStatement("DELETE FROM drug_inventory WHERE drug_id = ?", array($drug_id));
+    sqlStatement("DELETE FROM drug_templates WHERE drug_id = ?", array($drug_id));
+    sqlStatement("DELETE FROM drugs WHERE drug_id = ?", array($drug_id));
+    sqlStatement("DELETE FROM prices WHERE pr_id = ? AND pr_selector != ''", array($drug_id));
+  }
+
+  public Function updateDrug($data, $drug_id)
```
Comment (bradymiller) [inline services/DrugService.php]: do not capitalize function and fix this below also

File: interface/drugs/add_edit_drug.php [hunk 11]
```diff
@@ -176,55 +182,50 @@ function sel_related() {
...
-        ")");
+           $addDrug = array(
+            "name" => escapedff('form_name'),
+            "ndc_number" => escapedff('form_ndc_number'),
+            "drug_code" => escapedff('form_drug_code'),
+            "on_order" => escapedff('form_on_order'),
+            "reorder_point" => escapedff('form_reorder_point'),
+            "max_level" => escapedff('form_max_level'),
+            "form" => escapedff('form_form'),
+            "size" => escapedff('form_sixe'),
```
Comment (bradymiller) [inline interface/drugs/add_edit_drug.php]: typo

**Comment by bradymiller** [comment]
review completed. looking forward to next revision.

**Comment by robertogagliotta** [comment]
Hi Brady
I ll be back into this PR in few days. 
Thanks

**Comment by bradymiller** [comment]
hi @robertogagliotta , Looking forward to seeing the next revision! -brady

**Comment by robertogagliotta** [comment]
Hi @bradymiller 
During my unit testing I m not able to dispense a drug to a user I have "Inventory is  not available for this order"
I see a problem when adding a drug to the inventory, the On Hand field is 0 all the times.
Is this a Bug? please advice.
Do you want me to research deeper in this issue?
I'm doing something wrong?

**Comment by bradymiller** [comment]
hi @robertogagliotta ,
Is this happening in the development codebase (can test this on the development demo here: http://www.open-emr.org/wiki/index.php/Development_Demo#Without_demo_data ).
-brady

**Comment by robertogagliotta** [comment]
Tested on Dev Demo wo data
inventory1
inventory2
inventory3

**Comment by robertogagliotta** [comment]
The same "Inventory not available" warning on 5.0 patch 5

**Comment by MatthewVita** [comment]
@robertogagliotta Code is looking good so far! Can you provide the Apache error log for this warning?

**Comment by robertogagliotta** [comment]
Hi @MatthewVita 
https://github.com/openemr/openemr/blob/master/interface/drugs/dispense_drug.php#L40-L81
openemr code in Line 61 is responsable for that error message.

**Comment by MatthewVita** [comment]
@robertogagliotta Hi Roberto. Can you please either set debugging breakpoints or echo statements in this function to see where it is failing? https://github.com/openemr/openemr/blob/33ef407f6364a6eb1f75d7b0172708f84775dd25/interface/drugs/drugs.inc.php#L44-L290

I wonder if it is trying to send an email and it's not setup on your system...

**Comment by robertogagliotta** [comment]
Line 86 $dispensable is 1 on drugs table by default. No $sale_id is return -> error message.

I changed field dispense manually to 0 on drug table and no error message and the drug sale is done.

If this is the problem I do not see how users can set it up the field dispense to 0= pharmacy on the UI.

**Comment by MatthewVita** [comment]
@robertogagliotta It looks like the default is 1.

```
ADD dispensable tinyint(1) NOT NULL DEFAULT 1 COMMENT '0 = pharmacy elsewhere, 1 = dispensed here';
```

However, there is `INSERT` logic below here that appears to return a sales id. Can you keep debugging to see how far the code goes?

**Comment by robertogagliotta** [comment]
Yes, dispensable in drugs table is 1 = dispensed here. I was not able to find the way in the UI to set on_hand field in drug_inventory table to the initial amount of pills in the inventory, when dispense the prescription the error message shows up. I set the on_hand field manually in the drug inventory table and the on_hand inventory field is update when the prescription is dispense; Pharmacy Dispensary Module works good after setting up the on_hand field manually in the drug_inventory table. 
The problem is that add_edit_lot.php does not allow to edit/enter the on_hand number in the UI.

**Comment by robertogagliotta** [comment]
https://github.com/openemr/openemr/blob/33ef407f6364a6eb1f75d7b0172708f84775dd25/interface/drugs/add_edit_lot.php#L384-L416

**Comment by robertogagliotta** [comment]
@MatthewVita  I'll go back to the Modernization of drugs. Let me know if you need more inventory debugging.

**Comment by MatthewVita** [comment]
Sounds good @robertogagliotta. I wonder if the users that take advantage of this feature edit the code/UI by hand to get around what you have correctly noted: " add_edit_lot.php does not allow to edit/enter the on_hand number in the UI."

Anyways, keep up the great work!

**Pull request closed without merging**